modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/InlineUtils.java

```diff
@@ -0,0 +1,186 @@
+     */
+    static int binaryTupleInlineSize(IndexDescriptor indexDescriptor) {
+        List<? extends ColumnDescriptor> columns = indexDescriptor.columns();
+
+        assert !columns.isEmpty();
+
+        boolean hasNullColumns = columns.stream().anyMatch(ColumnDescriptor::nullable);
+
+        // Let's calculate the inline size for all columns.
+        int columnsInlineSize = columns.stream().map(ColumnDescriptor::type).mapToInt(InlineUtils::inlineSize).sum();
+
+        int inlineSize = BinaryTupleCommon.HEADER_SIZE
+                + (hasNullColumns ? BinaryTupleCommon.nullMapSize(columns.size()) : 0)
+                + columns.size() * Math.min(MAX_BINARY_TUPLE_OFFSET_TABLE_ENTRY_SIZE, valueSizeToEntrySize(columnsInlineSize))
+                + columnsInlineSize;
+
+        return Math.min(inlineSize, MAX_BINARY_TUPLE_INLINE_SIZE);
+    }
+
+    /**
+     * Calculates the inline size of {@link BinaryTuple} that will be stored in the {@link BplusInnerIo} and {@link BplusLeafIo} item.
+     *
+     * @param pageSize Page size in bytes.
+     * @param itemHeaderSize Size of the item header that is stored in the {@link BplusInnerIo} and {@link BplusLeafIo}, in bytes.
+     * @param indexDescriptor Index descriptor.
+     * @return Inline size in bytes, no more than {@link #MAX_BINARY_TUPLE_INLINE_SIZE}.
+     */
+    public static int binaryTupleInlineSize(int pageSize, int itemHeaderSize, IndexDescriptor indexDescriptor) {
+        int maxInnerNodeItemSize = ((innerNodePayloadSize(pageSize) - CHILD_LINK_SIZE) / MIN_INNER_PAGE_ITEM_COUNT) - CHILD_LINK_SIZE;
+
+        int binaryTupleInlineSize = Math.min(maxInnerNodeItemSize - itemHeaderSize, binaryTupleInlineSize(indexDescriptor));
+
+        if (binaryTupleInlineSize >= MAX_BINARY_TUPLE_INLINE_SIZE) {
+            return MAX_BINARY_TUPLE_INLINE_SIZE;
+        }
+
+        // If we have variable length columns and there is enough unused space in the innerNodes and leafNodes, then we can try increasing
+        // the inline size for the BinaryTuple. Example: pageSize = 1024, binaryTupleInlineSize = 124, BplusLeafIo.HEADER_SIZE = 56.
+        // In this case, the innerNode will fit 7 (with child links) items and 7 items in the leafNode, while 52 bytes will remain free for
+        // the innerNode items, and 100 bytes for the leafNode items, which we could use. For a innerNode, we can add (52 / 7) = 7 bytes
+        // for each item (with link), and for a leafNode, (100 / 7) = 14 bytes for each item, so we can safely use the 7 extra bytes for the
+        // innerNode and leafNode per item.
+
+        if (indexDescriptor.columns().stream().anyMatch(c -> !c.type().spec().fixedLength())) {
+            int itemSize = binaryTupleInlineSize + itemHeaderSize;
+
+            // Let's calculate whether to remain a place in the BplusInnerIo and the BplusLeafIo.
+            int remainingInnerNodePayloadSize = (innerNodePayloadSize(pageSize) - CHILD_LINK_SIZE) % (itemSize + CHILD_LINK_SIZE);
```

Comment:
Let's simplify these calculations.
You already have a desired item size 'itemSize' 
`(itemSize = binaryTupleInlineSize + itemHeaderSize)`.

Then you can get amount of items that fit in the page, like 'innerNodeItemCount' below.
`int innerNodeItemCount = (innerNodePayloadSize(pageSize) - CHILD_LINK_SIZE) / (itemSize + CHILD_LINK_SIZE);`

Then you can just get available space for the every single item.
`int innerNodeItemMaxSize = (innerNodePayloadSize(pageSize) - CHILD_LINK_SIZE) / innerNodeItemCount;`

Values <='innerNodeItemMaxSize' can be safely used instead of 'itemSize' parameter.